unison-core/src/Unison/Name.hs

```diff
@@ -103,20 +118,26 @@ stripNamePrefix prefix@Name{} name@Name{} =
   Name <$> Text.stripPrefix (toText prefix <> mid) (toText name)
   where
   mid = if toText prefix == "." then "" else "."
-stripNamePrefix (Name' prefix) (Name' name) =
-  fmap Name' do
-    suffix <- List.stripPrefix (toList prefix) (toList name)
-    NonEmpty.nonEmpty suffix
+stripNamePrefix n1@(Name' p1 prefix) n2@(Name' p2 name) =
+  if p1 == p2 then
+    fmap (Name' Relative) do
+      suffix <- List.stripPrefix (toList prefix) (toList name)
+      NonEmpty.nonEmpty suffix
+  else
+    error ("stripNamePrefix (" ++ show n1 ++ ") (" ++ show n2 ++ ")")
 
 joinDot :: Name -> Name -> Name
 joinDot prefix@Name{} suffix@Name{} =
   if toText prefix == "." then Name (toText prefix <> toText suffix)
   else Name (toText prefix <> "." <> toText suffix)
-joinDot (Name' prefix) (Name' suffix) = Name' (prefix <> suffix)
+joinDot n1@(Name' p1 prefix) n2@(Name' p2 suffix) =
+  case p2 of
+    Absolute -> error ("joinDot (" ++ show n1 ++ ") (" ++ show n2 ++ ")")
+    Relative -> Name' p1 (prefix <> suffix)
 
 unqualified :: Name -> Name
 unqualified (Name name) = unsafeFromText (last (Text.splitOn "." name))
-unqualified (Name' names) = Name' (NonEmpty.last names :| [])
+unqualified (Name' _ names) = Name' Relative (NonEmpty.last names :| [])
```

Comment:
This function's existence seems fishy to me; like an intermediate step in something more meaningful.

Reply:
Agreed